File: src/algorithms/trees/bst-operations/bst-validation-iterative/sources/bst-validation-iterative_test.py
import importlib
import sys
import os

sys.path.insert(0, os.path.dirname(__file__))
module = importlib.import_module("bst-validation-iterative")
BSTNode = module.BSTNode
bst_validation_iterative = module.bst_validation_iterative


def make_node(value, left=None, right=None):
    node = BSTNode(value)
    node.left = left
    node.right = right
    return node


def test_validates_correct_bst():
    tree = make_node(4, make_node(2, make_node(1), make_node(3)), make_node(6, make_node(5), make_node(7)))
    assert bst_validation_iterative(tree) == True


def test_rejects_invalid_bst():
    invalid = make_node(5, make_node(6), make_node(7))
    assert bst_validation_iterative(invalid) == False


def test_accepts_null():
    assert bst_validation_iterative(None) == True


def test_accepts_single_node():
    assert bst_validation_iterative(make_node(10)) == True


if __name__ == "__main__":
    test_validates_correct_bst()
    test_rejects_invalid_bst()
    test_accepts_null()
    test_accepts_single_node()
    print("All tests passed!")

File: src/algorithms/trees/bst-operations/bst-validation-iterative/__tests__/bst-validation-iterative.test.ts
import { describe, it, expect } from "vitest";
import { bstValidationIterative } from "../sources/bst-validation-iterative.ts?fn";

interface BSTNode {
  value: number;
  left: BSTNode | null;
  right: BSTNode | null;
}
function node(value: number, left: BSTNode | null = null, right: BSTNode | null = null): BSTNode {
  return { value, left, right };
}

describe("bstValidationIterative", () => {
  it("validates a correct BST", () => {
    const tree = node(4, node(2, node(1), node(3)), node(6, node(5), node(7)));
    expect(bstValidationIterative(tree)).toBe(true);
  });
  it("rejects an invalid BST", () => {
    const invalid = node(5, node(6), node(7));
    expect(bstValidationIterative(invalid)).toBe(false);
  });
  it("accepts null", () => {
    expect(bstValidationIterative(null)).toBe(true);
  });
  it("accepts single node", () => {
    expect(bstValidationIterative(node(10))).toBe(true);
  });
});
